craft_parts/utils/partition_utils.py

```diff
@@ -118,58 +115,123 @@ def _validate_partition_naming_convention(partitions: Sequence[str]) -> None:
+        - `foo` and 'foo/bar' conflicts
+        - `foo/bar` and 'foo/bar/baz' conflicts
+      2: environment variables derived from partition names must not conflicts
+        - `foo/bar-baz` and `foo/bar/baz` conflicts (would be converted to FOO_BAR_BAZ)
+        - `foo/bar/baz-qux` and `foo/bar-baz/qux` conflicts (would be converted to FOO_BAR_BAZ_QUX)
 
-    :raises FeatureError: for namespace conflicts
     """
-    namespaced_partitions: set[str] = set()
-    regular_partitions: set[str] = set()
+    conflictings_partitions: dict[str, set[str]] = defaultdict(set)
+
+    # Rule #1: Detect conflicts between partitions and namespaces partitions
+    for candidate_partition in partitions:
+        for partition in partitions:
+            if candidate_partition == partition:
+                continue
+            if namespace_conflicts(candidate_partition, partition):
+                if not conflictings_partitions[partition]:
+                    conflictings_partitions[partition].add(partition)
+                conflictings_partitions[partition].add(candidate_partition)
+
+    # Rule #2: Detect conflicts between environment variable names
+    underscored_partitions: dict[str, str] = {}
 
-    # sort partitions
     for partition in partitions:
-        if _is_valid_partition_name(partition):
-            regular_partitions.add(partition)
+        underscored = partition.translate({ord("-"): "_", ord("/"): "_"})
+
+        for known_partition, known_underscored in underscored_partitions.items():
+            if env_var_conflicts(underscored, known_underscored):
+                if not conflictings_partitions[known_partition]:
+                    conflictings_partitions[known_partition].add(known_partition)
+                conflictings_partitions[known_partition].add(partition)
+
+        underscored_partitions[partition] = underscored
+
+    return conflictings_partitions
+
+
+def split_and_detect(a: str, b: str, separator: str) -> bool:
+    """Split candidates with the separator and check if the shorter one is a subset of the other."""
+    a_list = a.split(separator)
+    b_list = b.split(separator)
+
+    if len(a_list) > len(b_list):
+        return a_list[: len(b_list)] == b_list
+
+    return b_list[: len(a_list)] == a_list
+
+
+def namespace_conflicts(a: str, b: str) -> bool:
+    """Detect conflicts between partitions and namespaces."""
+    return split_and_detect(a, b, "/")
+
+
+def env_var_conflicts(a: str, b: str) -> bool:
+    """Detect conflicts between partitions when transformed into environment variables."""
+    return split_and_detect(a, b, "_")
+
+
+def unique_conflicts(
+    conflicting_partitions: dict[str, set[str]],
+) -> dict[str, set[str]]:
+    """Remove subsets of other conflicts."""
+    filtered: dict[str, set[str]] = {}
+    for partition, to_keep in conflicting_partitions.items():
+        for p2, conflicts in conflicting_partitions.items():
```

Comment:
I wonder if there's a simpler way to do this. Maybe would could just create a set of conflicting names as they are detected? I'm not sure what you are returning here, a more descriptive docstring including the return value would help.

Reply:
I reworked this section based on your suggestion. I have now added a proper `register_conflict()` function to avoid registering duplicates (and so avoid having to filter them later).
It now looks a lot cleaner and easier to read.